File: sdk/src/profiles/geo/apply.js
// weavepack-geo — delta application (applyChain).
//
// Live collection state:
//   {
//     name:    string,
//     fidKind: number,   // FID_KIND constant; -1 = not yet established
//     features: LiveFeature[],
//   }
//
// LiveFeature:
//   {
//     fid:     string | BigInt | null,
//     geomType:       number,
//     coordPrecision: number,
//     hasZ:           boolean,
//     geom:           object (same shape as encoder geom sections),
//     props:          Map<string, { ctype, value }>,
//   }
//
// Profile isolation: imports only from ./types.js.

import { FID_KIND, OP, PATH_KIND } from "./types.js"

// ── Helpers ────────────────────────────────────────────────────────────────

function cloneState(state) {
  return {
    name:     state.name,
    fidKind:  state.fidKind,
    features: state.features.map(f => ({
      fid:            f.fid,
      geomType:       f.geomType,
      coordPrecision: f.coordPrecision,
      hasZ:           f.hasZ,
      geom:           f.geom,
      props:          new Map(f.props),
    })),
  }
}

// Extract per-feature data from a decoded feature_block or gc_block.
function featuresFromBlock(blk) {
  const features = []
  const n = blk.numFeatures

  if (blk.type === 'feature') {
    for (let i = 0; i < n; i++) {
      const fid = blk.fids ? blk.fids[i] : null
      const geom = extractSingleFeatureGeom(blk, i)
      const props = new Map()
      for (const col of (blk.propCols ?? [])) {
        if (col.values[i] !== null && col.values[i] !== undefined) {
          props.set(col.name, { ctype: col.ctype, value: col.values[i] })
        }
      }
      features.push({
        fid,
        geomType:       blk.geomType,
        coordPrecision: blk.coordPrecision ?? 0,
        hasZ:           blk.hasZ ?? false,
        geom,
        props,
      })
    }
  } else if (blk.type === 'geometry_collection') {
    for (let i = 0; i < n; i++) {
      const fid = blk.fids ? blk.fids[i] : null
      const props = new Map()
      for (const col of (blk.propCols ?? [])) {
        if (col.values[i] !== null && col.values[i] !== undefined) {
          props.set(col.name, { ctype: col.ctype, value: col.values[i] })
        }
      }
      features.push({
        fid,
        geomType:       6, // GEOMETRY_COLLECTION
        coordPrecision: blk.coordPrecision ?? 0,
        hasZ:           blk.hasZ ?? false,
        geom:           { type: 'gc', subGeomCounts: blk.subGeomCounts, subGeoms: blk.subGeoms },
        props,
      })
    }
  }

  return features
}

// Slice the columnar geom section to get a single feature's geometry.
function extractSingleFeatureGeom(blk, i) {
  const g = blk.geom ?? {}
  switch (blk.geomType) {
    case 0: // POINT
      return {
        xCol: [g.xCol[i]],
        yCol: [g.yCol[i]],
        ...(blk.hasZ ? { zCol: [g.zCol[i]] } : {}),
      }
    case 1: { // LINESTRING
      const count  = g.coordCounts[i]
      const offset = g.coordCounts.slice(0, i).reduce((s, c) => s + c, 0)
      return {
        coordCounts: [count],
        xCol: g.xCol.slice(offset, offset + count),
        yCol: g.yCol.slice(offset, offset + count),
        ...(blk.hasZ ? { zCol: g.zCol.slice(offset, offset + count) } : {}),
      }
    }
    case 2: { // POLYGON
      const ringsForFeature = g.ringsPerFeature[i]
      const ringOffset      = g.ringsPerFeature.slice(0, i).reduce((s, c) => s + c, 0)
      const myRingCounts    = g.ringCounts.slice(ringOffset, ringOffset + ringsForFeature)
      const vertexOffset    = g.ringCounts.slice(0, ringOffset).reduce((s, c) => s + c, 0)
      const totalV          = myRingCounts.reduce((s, c) => s + c, 0)
      return {
        ringsPerFeature: [ringsForFeature],
        ringCounts:      myRingCounts,
        xCol: g.xCol.slice(vertexOffset, vertexOffset + totalV),
        yCol: g.yCol.slice(vertexOffset, vertexOffset + totalV),
        ...(blk.hasZ ? { zCol: g.zCol.slice(vertexOffset, vertexOffset + totalV) } : {}),
      }
    }
    case 3: { // MULTIPOINT
      const count  = g.partCounts[i]
      const offset = g.partCounts.slice(0, i).reduce((s, c) => s + c, 0)
      return {
        partCounts: [count],
        xCol: g.xCol.slice(offset, offset + count),
        yCol: g.yCol.slice(offset, offset + count),
        ...(blk.hasZ ? { zCol: g.zCol.slice(offset, offset + count) } : {}),
      }
    }
    case 4: { // MULTILINESTRING
      const lineCount  = g.partCounts[i]
      const lineOffset = g.partCounts.slice(0, i).reduce((s, c) => s + c, 0)
      const myCoordCounts  = g.coordCounts.slice(lineOffset, lineOffset + lineCount)
      const vertexOffset   = g.coordCounts.slice(0, lineOffset).reduce((s, c) => s + c, 0)
      const totalV         = myCoordCounts.reduce((s, c) => s + c, 0)
      return {
        partCounts:  [lineCount],
        coordCounts: myCoordCounts,
        xCol: g.xCol.slice(vertexOffset, vertexOffset + totalV),
        yCol: g.yCol.slice(vertexOffset, vertexOffset + totalV),
        ...(blk.hasZ ? { zCol: g.zCol.slice(vertexOffset, vertexOffset + totalV) } : {}),
      }
    }
    case 5: { // MULTIPOLYGON
      const polyCount  = g.partCounts[i]
      const polyOffset = g.partCounts.slice(0, i).reduce((s, c) => s + c, 0)
      const myRingsPerPart = g.ringsPerPart.slice(polyOffset, polyOffset + polyCount)
      const ringOffset     = g.ringsPerPart.slice(0, polyOffset).reduce((s, c) => s + c, 0)
      const totalRings     = myRingsPerPart.reduce((s, c) => s + c, 0)
      const myRingCounts   = g.ringCounts.slice(ringOffset, ringOffset + totalRings)
      const vertexOffset   = g.ringCounts.slice(0, ringOffset).reduce((s, c) => s + c, 0)
      const totalV         = myRingCounts.reduce((s, c) => s + c, 0)
      return {
        partCounts:  [polyCount],
        ringsPerPart: myRingsPerPart,
        ringCounts:  myRingCounts,
        xCol: g.xCol.slice(vertexOffset, vertexOffset + totalV),
        yCol: g.yCol.slice(vertexOffset, vertexOffset + totalV),
        ...(blk.hasZ ? { zCol: g.zCol.slice(vertexOffset, vertexOffset + totalV) } : {}),
      }
    }
    case 7: // NULL_GEOMETRY
      return {}
    default:
      return g
  }
}

// ── Feature lookup ─────────────────────────────────────────────────────────

function resolveInnerPath(state, path) {
  if (path.kind === PATH_KIND.FEAT_BY_IDX) {
    if (path.index >= state.features.length)
      throw new Error(`feature_index_out_of_bounds: ${path.index}`)
    return path.index
  }
  if (path.kind === PATH_KIND.FEAT_BY_STR_FID) {
    if (state.fidKind !== FID_KIND.FID_STRING)
      throw new Error("fid_kind_mismatch")
    const idx = state.features.findIndex(f => f.fid === path.fid)
    if (idx < 0) throw new Error(`feature_not_found: ${path.fid}`)
    return idx
  }
  if (path.kind === PATH_KIND.FEAT_BY_INT_FID) {
    if (state.fidKind !== FID_KIND.FID_UINT64)
      throw new Error("fid_kind_mismatch")
    const target = BigInt(path.fid)
    const idx = state.features.findIndex(f => BigInt(f.fid) === target)
    if (idx < 0) throw new Error(`feature_not_found: ${path.fid}`)
    return idx
  }
  throw new Error(`expected feature path, got kind ${path.kind}`)
}

// ── Op application ─────────────────────────────────────────────────────────

function applyOp(state, op) {
  state = cloneState(state)

  switch (op.op) {
    case OP.FEATURE_INSERT: {
      const inserted = featuresFromBlock(op.block)
      const newFidKind = op.block.fidKind ?? FID_KIND.FID_ABSENT
      if (state.fidKind === -1) {
        state.fidKind = newFidKind
      } else if (state.fidKind !== newFidKind) {
        throw new Error("fid_kind_mismatch")
      }
      if (newFidKind !== FID_KIND.FID_ABSENT) {
        const existingFids = new Set()
        if (newFidKind === FID_KIND.FID_UINT64) {
          for (const f of state.features) existingFids.add(String(BigInt(f.fid)))
          for (const f of inserted) {
            const k = String(BigInt(f.fid))
            if (existingFids.has(k)) throw new Error("duplicate_fid")
            existingFids.add(k)
          }
        } else {
          for (const f of state.features) existingFids.add(f.fid)
          for (const f of inserted) {
            if (existingFids.has(f.fid)) throw new Error("duplicate_fid")
            existingFids.add(f.fid)
          }
        }
      }
      state.features.push(...inserted)
      break
    }

    case OP.FEATURE_DELETE: {
      if (op.mode === 0) {
        // path_list — delete in descending index order to avoid shift issues
        const indices = op.paths.map(p => resolveInnerPath(state, p))
        const seen = new Set()
        for (const idx of indices) {
          if (seen.has(idx)) throw new Error("duplicate_delete_target")
          seen.add(idx)
        }
        indices.sort((a, b) => b - a)
        for (const idx of indices) state.features.splice(idx, 1)
      } else if (op.mode === 1) {
        if (op.start + op.count > state.features.length)
          throw new Error("feature_index_out_of_bounds")
        state.features.splice(op.start, op.count)
      }
      break
    }

    case OP.GEOMETRY_REPLACE: {
      const path  = op.path
      if (path.kind !== PATH_KIND.FEAT_GEOMETRY)
        throw new Error(`geometry_replace: expected FEAT_GEOMETRY path, got kind ${path.kind}`)
      const idx   = resolveInnerPath(state, path.inner)
      const feat  = state.features[idx]
      const newBlk = op.block
      const newGeom = extractSingleFeatureGeom({ ...newBlk, numFeatures: 1 }, 0)
      state.features[idx] = {
        ...feat,
        geomType:       newBlk.geomType,
        coordPrecision: newBlk.coordPrecision ?? feat.coordPrecision,
        hasZ:           newBlk.hasZ ?? feat.hasZ,
        geom:           newGeom,
      }
      break
    }

    case OP.PROP_SET: {
      const path = op.path
      let featIdx, propKey
      if (path.kind === PATH_KIND.FEAT_PROP_NAME) {
        featIdx = resolveInnerPath(state, path.inner)
        propKey = path.name
      } else if (path.kind === PATH_KIND.FEAT_PROP_IDX) {
        featIdx = resolveInnerPath(state, path.inner)
        // Translate column index to property name via existing props order
        const keys = [...state.features[featIdx].props.keys()]
        if (path.colIdx >= keys.length) throw new Error("col_idx_out_of_bounds")
        propKey = keys[path.colIdx]
      } else {
        throw new Error(`prop_set: unsupported path kind ${path.kind}`)
      }
      state.features[featIdx].props.set(propKey, { ctype: op.ctype, value: op.value })
      break
    }

    case OP.PROP_DELETE: {
      const path = op.path
      let featIdx, propKey
      if (path.kind === PATH_KIND.FEAT_PROP_NAME) {
        featIdx = resolveInnerPath(state, path.inner)
        propKey = path.name
      } else if (path.kind === PATH_KIND.FEAT_PROP_IDX) {
        featIdx = resolveInnerPath(state, path.inner)
        const keys = [...state.features[featIdx].props.keys()]
        if (path.colIdx >= keys.length) throw new Error("col_idx_out_of_bounds")
        propKey = keys[path.colIdx]
      } else {
        throw new Error(`prop_delete: unsupported path kind ${path.kind}`)
      }
      // idempotent: no error if property absent
      state.features[featIdx].props.delete(propKey)
      break
    }

    case OP.COLLECTION_REPLACE: {
      state.features = []
      state.fidKind  = -1
      for (const blk of (op.blocks ?? [])) {
        const inserted = featuresFromBlock(blk)
        const newFidKind = blk.fidKind ?? FID_KIND.FID_ABSENT
        if (state.fidKind === -1) state.fidKind = newFidKind
        else if (state.fidKind !== newFidKind) throw new Error("fid_kind_mismatch")
        state.features.push(...inserted)
      }
      if (state.fidKind === -1) state.fidKind = FID_KIND.FID_ABSENT
      break
    }

    default:
      throw new Error(`unknown_delta_op: ${op.op}`)
  }

  return state
}

// ── Public ─────────────────────────────────────────────────────────────────

export function initState(doc) {
  const state = {
    name:     doc.name ?? "",
    fidKind:  -1,
    features: [],
  }
  for (const blk of (doc.blocks ?? [])) {
    if (blk.type === 'delta') continue  // snapshots only
    const inserted = featuresFromBlock(blk)
    const newFidKind = blk.fidKind ?? FID_KIND.FID_ABSENT
    if (state.fidKind === -1) state.fidKind = newFidKind
    else if (state.fidKind !== newFidKind) throw new Error("fid_kind_mismatch")
    state.features.push(...inserted)
  }
  if (state.fidKind === -1) state.fidKind = FID_KIND.FID_ABSENT
  return state
}

export function applyChain(state, ops) {
  let s = state
  for (const op of ops) s = applyOp(s, op)
  return s
}

File: impl/rust/weavepack-geo/src/apply.rs
// weavepack-geo — delta application.
//
// Mirrors sdk/src/profiles/geo/apply.js.
// Profile isolation: only imports from crate::types.

use crate::types::{
    Block, CellValue, DocBlock, FeatureBlock, Fid, GeoDocument, Geom,
    InnerPath, Op, Path,
    FID_ABSENT, FID_STRING, FID_UINT64,
    GEOM_LINESTRING, GEOM_MULTILINESTRING, GEOM_MULTIPOINT, GEOM_MULTIPOLYGON,
    GEOM_NULL, GEOM_POINT, GEOM_POLYGON,
};

// ── Live property entry ───────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct LiveProp {
    pub ctype: u8,
    pub value: CellValue,
}

// ── Live feature ────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct LiveFeature {
    pub fid:             Option<Fid>,
    pub geom_type:       u8,
    pub coord_precision: u8,
    pub has_z:           bool,
    pub geom:            Geom,
    pub props:           Vec<(String, LiveProp)>,  // insertion-ordered
}

// ── Live state ─────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct GeoState {
    pub name:     String,
    pub fid_kind: i8,   // -1 = unset, 0/1/2 = established
    pub features: Vec<LiveFeature>,
}

// ── Geometry slicing (columnar → per-feature) ────────────────────────────────

fn sum_u32(arr: &[u32]) -> usize { arr.iter().map(|&v| v as usize).sum() }

fn slice_f64(v: &[f64], start: usize, len: usize) -> Vec<f64> {
    v[start..start + len].to_vec()
}

fn extract_single_feature_geom(blk: &FeatureBlock, i: usize) -> Geom {
    let g = &blk.geom;
    match blk.geom_type {
        GEOM_POINT => Geom {
            x: vec![g.x[i]],
            y: vec![g.y[i]],
            z: blk.has_z.then(|| vec![g.z.as_ref().unwrap()[i]]),
            ..Default::default()
        },
        GEOM_LINESTRING => {
            let cnt    = g.coord_counts[i] as usize;
            let offset = sum_u32(&g.coord_counts[..i]);
            Geom {
                coord_counts: vec![g.coord_counts[i]],
                x: slice_f64(&g.x, offset, cnt),
                y: slice_f64(&g.y, offset, cnt),
                z: blk.has_z.then(|| slice_f64(g.z.as_deref().unwrap(), offset, cnt)),
                ..Default::default()
            }
        }
        GEOM_POLYGON => {
            let rings_for_feat = g.rings_per_feature[i] as usize;
            let ring_offset    = sum_u32(&g.rings_per_feature[..i]);
            let my_ring_counts = g.ring_counts[ring_offset..ring_offset + rings_for_feat].to_vec();
            let vertex_offset  = sum_u32(&g.ring_counts[..ring_offset]);
            let total_v        = sum_u32(&my_ring_counts);
            Geom {
                rings_per_feature: vec![rings_for_feat as u32],
                ring_counts: my_ring_counts,
                x: slice_f64(&g.x, vertex_offset, total_v),
                y: slice_f64(&g.y, vertex_offset, total_v),
                z: blk.has_z.then(|| slice_f64(g.z.as_deref().unwrap(), vertex_offset, total_v)),
                ..Default::default()
            }
        }
        GEOM_MULTIPOINT => {
            let cnt    = g.part_counts[i] as usize;
            let offset = sum_u32(&g.part_counts[..i]);
            Geom {
                part_counts: vec![g.part_counts[i]],
                x: slice_f64(&g.x, offset, cnt),
                y: slice_f64(&g.y, offset, cnt),
                z: blk.has_z.then(|| slice_f64(g.z.as_deref().unwrap(), offset, cnt)),
                ..Default::default()
            }
        }
        GEOM_MULTILINESTRING => {
            let line_cnt    = g.part_counts[i] as usize;
            let line_offset = sum_u32(&g.part_counts[..i]);
            let my_coord_counts = g.coord_counts[line_offset..line_offset + line_cnt].to_vec();
            let vertex_offset   = sum_u32(&g.coord_counts[..line_offset]);
            let total_v         = sum_u32(&my_coord_counts);
            Geom {
                part_counts:  vec![g.part_counts[i]],
                coord_counts: my_coord_counts,
                x: slice_f64(&g.x, vertex_offset, total_v),
                y: slice_f64(&g.y, vertex_offset, total_v),
                z: blk.has_z.then(|| slice_f64(g.z.as_deref().unwrap(), vertex_offset, total_v)),
                ..Default::default()
            }
        }
        GEOM_MULTIPOLYGON => {
            let poly_cnt    = g.part_counts[i] as usize;
            let poly_offset = sum_u32(&g.part_counts[..i]);
            let my_rings_per_part = g.rings_per_part[poly_offset..poly_offset + poly_cnt].to_vec();
            let ring_offset       = sum_u32(&g.rings_per_part[..poly_offset]);
            let total_rings       = sum_u32(&my_rings_per_part);
            let my_ring_counts    = g.ring_counts[ring_offset..ring_offset + total_rings].to_vec();
            let vertex_offset     = sum_u32(&g.ring_counts[..ring_offset]);
            let total_v           = sum_u32(&my_ring_counts);
            Geom {
                part_counts:   vec![g.part_counts[i]],
                rings_per_part: my_rings_per_part,
                ring_counts:   my_ring_counts,
                x: slice_f64(&g.x, vertex_offset, total_v),
                y: slice_f64(&g.y, vertex_offset, total_v),
                z: blk.has_z.then(|| slice_f64(g.z.as_deref().unwrap(), vertex_offset, total_v)),
                ..Default::default()
            }
        }
        GEOM_NULL | _ => Geom::default(),
    }
}

// ── Block → Vec<LiveFeature> ────────────────────────────────────────────────────

fn features_from_block(block: &Block) -> Result<(Vec<LiveFeature>, u8), String> {
    match block {
        Block::Feature(blk) => {
            let n = blk.num_features;
            let mut features = Vec::with_capacity(n);
            for i in 0..n {
                let fid = blk.fids.as_ref().map(|fids| fids[i].clone());
                let geom = extract_single_feature_geom(blk, i);
                let mut props: Vec<(String, LiveProp)> = Vec::new();
                for col in &blk.prop_cols {
                    if let Some(val) = &col.values[i] {
                        props.push((col.name.clone(), LiveProp { ctype: col.ctype, value: val.clone() }));
                    }
                }
                features.push(LiveFeature {
                    fid,
                    geom_type:       blk.geom_type,
                    coord_precision: blk.coord_precision,
                    has_z:           blk.has_z,
                    geom,
                    props,
                });
            }
            Ok((features, blk.fid_kind))
        }
        Block::Gc(blk) => {
            let n = blk.num_features;
            let mut features = Vec::with_capacity(n);
            for i in 0..n {
                let fid = blk.fids.as_ref().map(|fids| fids[i].clone());
                let mut props: Vec<(String, LiveProp)> = Vec::new();
                for col in &blk.prop_cols {
                    if let Some(val) = &col.values[i] {
                        props.push((col.name.clone(), LiveProp { ctype: col.ctype, value: val.clone() }));
                    }
                }
                features.push(LiveFeature {
                    fid,
                    geom_type:       6, // GEOM_GEOMETRY_COLLECTION
                    coord_precision: blk.coord_precision,
                    has_z:           blk.has_z,
                    geom:            Geom::default(),
                    props,
                });
            }
            Ok((features, blk.fid_kind))
        }
    }
}

// ── Feature resolution ───────────────────────────────────────────────────────────

fn resolve_inner_path(state: &GeoState, ip: &InnerPath) -> Result<usize, String> {
    match ip {
        InnerPath::ByIdx(idx) => {
            let i = *idx as usize;
            if i >= state.features.len() {
                return Err(format!("feature_index_out_of_bounds: {idx}"));
            }
            Ok(i)
        }
        InnerPath::ByStrFid(fid) => {
            if state.fid_kind != FID_STRING as i8 {
                return Err("fid_kind_mismatch".into());
            }
            state.features.iter().position(|f| {
                matches!(&f.fid, Some(Fid::Str(s)) if s == fid)
            }).ok_or_else(|| format!("feature_not_found: {fid}"))
        }
        InnerPath::ByIntFid(fid) => {
            if state.fid_kind != FID_UINT64 as i8 {
                return Err("fid_kind_mismatch".into());
            }
            let target = *fid;
            state.features.iter().position(|f| {
                matches!(&f.fid, Some(Fid::Int(n)) if *n == target)
            }).ok_or_else(|| format!("feature_not_found: {fid}"))
        }
    }
}

// ── Insert helper ─────────────────────────────────────────────────────────────

fn insert_features(state: &mut GeoState, block: &Block) -> Result<(), String> {
    let (inserted, new_fk) = features_from_block(block)?;
    if state.fid_kind == -1 {
        state.fid_kind = new_fk as i8;
    } else if state.fid_kind != new_fk as i8 {
        return Err("fid_kind_mismatch".into());
    }
    // Duplicate FID check
    if new_fk != FID_ABSENT {
        for ins in &inserted {
            if let Some(ref fid) = ins.fid {
                let dup = if new_fk == FID_UINT64 {
                    let target = match fid { Fid::Int(n) => *n, _ => 0 };
                    state.features.iter().any(|f| matches!(&f.fid, Some(Fid::Int(n)) if *n == target))
                        || inserted.iter().take_while(|x| !std::ptr::eq(*x, ins)).any(|f|
                            matches!(&f.fid, Some(Fid::Int(n)) if *n == {match fid { Fid::Int(n) => *n, _ => 1 }}))
                } else {
                    let s = match fid { Fid::Str(s) => s.as_str(), _ => "" };
                    state.features.iter().any(|f| matches!(&f.fid, Some(Fid::Str(t)) if t == s))
                };
                if dup { return Err("duplicate_fid".into()); }
            }
        }
    }
    state.features.extend(inserted);
    Ok(())
}

// ── Op application ─────────────────────────────────────────────────────────────

fn apply_op(state: &mut GeoState, op: &Op) -> Result<(), String> {
    match op {
        Op::FeatureInsert { block } => {
            insert_features(state, block)?;
        }

        Op::FeatureDelete { mode, paths, start, count } => {
            if *mode == 0 {
                let mut indices: Vec<usize> = paths.iter()
                    .map(|p| match p {
                        Path::ByIdx(i)     => resolve_inner_path(state, &InnerPath::ByIdx(*i)),
                        Path::ByStrFid(s)  => resolve_inner_path(state, &InnerPath::ByStrFid(s.clone())),
                        Path::ByIntFid(n)  => resolve_inner_path(state, &InnerPath::ByIntFid(*n)),
                        _ => Err("feature_delete: expected direct feature path".into()),
                    })
                    .collect::<Result<_, _>>()?;
                indices.sort_unstable_by(|a, b| b.cmp(a));  // descending
                for idx in indices { state.features.remove(idx); }
            } else if *mode == 1 {
                let start = *start as usize;
                let count = *count as usize;
                if start + count > state.features.len() {
                    return Err("feature_index_out_of_bounds".into());
                }
                state.features.drain(start..start + count);
            } else {
                return Err(format!("unknown_feature_delete_mode: {mode}"));
            }
        }

        Op::GeometryReplace { path, block } => {
            let inner = match path {
                Path::Geometry(ip) => ip,
                _ => return Err("geometry_replace: expected FEAT_GEOMETRY path".into()),
            };
            let idx  = resolve_inner_path(state, inner)?;
            let feat = &mut state.features[idx];
            let dummy = FeatureBlock {
                geom_type:       block.geom_type,
                coord_precision: block.coord_precision,
                has_z:           block.has_z,
                fid_kind:        block.fid_kind,
                num_features:    1,
                fids:            None,
                geom:            block.geom.clone(),
                prop_cols:       Vec::new(),
            };
            let new_geom = extract_single_feature_geom(&dummy, 0);
            feat.geom_type       = block.geom_type;
            feat.coord_precision = block.coord_precision;
            feat.has_z           = block.has_z;
            feat.geom            = new_geom;
        }

        Op::PropSet { path, ctype, value } => {
            let (feat_idx, prop_key) = resolve_prop_path(state, path)?;
            let feat = &mut state.features[feat_idx];
            if let Some(e) = feat.props.iter_mut().find(|(k, _)| k == &prop_key) {
                e.1 = LiveProp { ctype: *ctype, value: value.clone() };
            } else {
                feat.props.push((prop_key, LiveProp { ctype: *ctype, value: value.clone() }));
            }
        }

        Op::PropDelete { path } => {
            let (feat_idx, prop_key) = resolve_prop_path(state, path)?;
            let feat = &mut state.features[feat_idx];
            feat.props.retain(|(k, _)| k != &prop_key);
        }

        Op::CollectionReplace { blocks } => {
            state.features.clear();
            state.fid_kind = -1;
            for blk in blocks {
                insert_features(state, blk)?;
            }
            if state.fid_kind == -1 { state.fid_kind = FID_ABSENT as i8; }
        }
    }
    Ok(())
}

fn resolve_prop_path(state: &GeoState, path: &Path) -> Result<(usize, String), String> {
    match path {
        Path::PropName { inner, name } => {
            let idx = resolve_inner_path(state, inner)?;
            Ok((idx, name.clone()))
        }
        Path::PropIdx { inner, col_idx } => {
            let idx  = resolve_inner_path(state, inner)?;
            let cidx = *col_idx as usize;
            let keys: Vec<&str> = state.features[idx].props.iter().map(|(k, _)| k.as_str()).collect();
            if cidx >= keys.len() {
                return Err("col_idx_out_of_bounds".into());
            }
            Ok((idx, keys[cidx].to_owned()))
        }
        _ => Err(format!("prop_set/delete: unsupported path kind")),
    }
}

// ── Public: init_state ──────────────────────────────────────────────────────────

pub fn init_state(doc: &GeoDocument) -> Result<GeoState, String> {
    let mut state = GeoState {
        name:     doc.name.clone(),
        fid_kind: -1,
        features: Vec::new(),
    };
    for blk in &doc.blocks {
        let block = match blk {
            DocBlock::Feature(b) => Block::Feature(b.clone()),
            DocBlock::Gc(b)      => Block::Gc(b.clone()),
            DocBlock::Delta(_)   => continue,
        };
        insert_features(&mut state, &block)?;
    }
    if state.fid_kind == -1 { state.fid_kind = FID_ABSENT as i8; }
    Ok(state)
}

// ── Public: apply_chain ──────────────────────────────────────────────────────────

pub fn apply_chain(mut state: GeoState, ops: &[Op]) -> Result<GeoState, String> {
    for op in ops { apply_op(&mut state, op)?; }
    Ok(state)
}
